File: src/UFB/Universal.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;

class Universal{
  private Universal(){}
  /**
   * Returns a {@code String} with added zeroes to its left side.
   * <p>Example:</p>
   * <pre><code>
   * manPadding("1", 4);
   * return "0001"
   * </code></pre>
   * @param  str  the input {@code String} to add left-side padding to
   * @param  i    the appropriate length of {@code str} after padding
   * @return      the {@code str} with left-side padding
   * @see         String#format(String, Object...)
  */
	public static String manPadding(final String str, final int i){
    return String.format(
      new StringBuilder("%").append(i).append("s").toString(),
      str
    ).replace(" ", "0");
  }
  /**
   * Returns a {@code boolean} that signifies if a given {@code char} is a digit.
   * <p>Example:</p>
   * <pre><code>
   * isDigit('0');
   * true
   * </code></pre>
   * @param  c    the input {@code char} to check
   * @param  i    the appropriate length of {@code str} after padding
   * @return      true if {@code c} is a digit, false otherwise
   * @see         Character#isDigit(char)
  */
  public static boolean isDigit(final char c){
    // BeCoz Character.isDigit has too much function overhead.
    return (c>47&&c<58);
  }
  /**
   * Returns an {@code int} parsed from the input {@code String}.
   * <p>Example:</p>
   * <pre><code>
   * toIntAbsolute("100");
   * 100
   * </code></pre>
   * @param  s    the input {@code String} to parse
   * @return      the {@code int} representation of {@code s}
   * @see         Integer#parseInt(String)
  */
  public static int toIntAbsolute(final String s){
		// BeCoz Integer#parseInt() is slow and try-catch is expensive.
    int result=0;
    for(final char c:s.toCharArray()){
      result+=c-48;
      result*=10;
    }
    return result/10;
  }
  /**
   * Returns a {@code String} representation of the input {@code String[]}.
   * <p>Each element in the array is separated by one whitespace character.</p>
   * <p>Example:</p>
   * <pre><code>
   * lineGen(new String[]{"haha", "ha"});
   * "haha ha"
   * </code></pre>
   * @param  line the input {@code String[]} to re-represent
   * @return      the {@code String} representation of {@code line}
   * @see         Arrays#toString(Object[])
  */
	public static String lineGen(final String[]line){
    final String out=Arrays.toString(line).replace(", ", " ");
		return out.substring(1, out.length()-1).concat("\n\n");
	}
  /**
   * Returns a formatted {@code String} error from the input.
   * <p>Example:</p>
   * <pre><code>
   * final String[] line={"unknownCommand", "0", "90"};
   * </code></pre>
   * <pre><code>
   * formatError(line, "Command", line[0], "Does Not Exist");
   * """Error: |
   *        Command: |
   *            \"unknownCommand\" Does Not Exist: |
   *                unknownCommand 0 90
   * """
   * </code></pre>
   * @param  line the input {@code String[]} that represents the whole line of code
   * @param  in   the input {@code String...} that elaborates the error even further
   * @return      the {@code String} representation of the error itself
  */
  public static String formatError(final String[] line, final String... in){
    if(in.length<4)
      return new StringBuilder("Error: |\n")
            .append(String.format("    %s: |\n", in[0]))
            .append(String.format("        \"%s\" %s: |\n", in[1], in[2]))
            .append(String.format("            %s", lineGen(line))).toString();
    else
      return new StringBuilder("Error: |\n")
            .append(String.format("    %s: |\n", in[0]))
            .append(String.format("        \"%s\" %s: |\n", in[1], in[2]))
            .append(String.format("            %s", convertUnicode(lineGen(line).replace("\n\n", "\n"))))
            .append(String.format("        %s: |\n", "Which Is When Converted"))
            .append(String.format("            %s", in[3])).toString();
  }
  /**
   * Returns a formatted {@code String} that converts all unicode points.
   * <p>Example:</p>
   * <pre><code>
   * convertUnicode("uu0032");
   * " "
   * </code></pre>
   * @param  in   the input {@code String} to be scouted of unicode points
   * @return      the {@code String} with all the unicode points converted
  */
  public static String convertUnicode(final String in){
    if(in.length()<6)return in;
    String temp=in;
    // Regex slow ._.
    for(int i=0;i<temp.length()-6;i++){
      if(temp.substring(i, i+2).toLowerCase().equals("uu")){
        boolean confirmed=true;
        for(int i2=i+2;i2<6;i2++)
          if(!isDigit(temp.charAt(i2)))confirmed=false;
        if(confirmed)
          temp=new StringBuilder(temp.substring(0, i))
            .append((char)toIntAbsolute(
              temp.substring(i+2, i+6)
            ))
            .append(temp.substring(i+6)).toString();
      }
    }
    return temp;
  }
  final static HashMap<Character, Integer> memMap=new HashMap<>(){{
		put(' ', 0); put('A', 1); put('B', 2); put('C', 3); put('D', 4);
    put('E', 5); put('F', 6); put('G', 7); put('H', 8); put('I', 9);
		put('J', 10); put('K', 11); put('L', 12); put('M', 13); put('N', 14);
    put('O', 15); put('P', 16); put('Q', 17); put('R', 18); put('S', 19);
    put('T', 20); put('U', 21); put('V', 22); put('W', 23); put('X', 24);
    put('Y', 25); put('Z', 26);
		put('0', 27); put('1', 28); put('2', 29); put('3', 30); put('4', 31);
    put('5', 32); put('6', 33); put('7', 34); put('8', 35); put('9', 36);
		put('\n', 37);
  }};
  /**
   * Returns a {@code String[]} full of UFB memory pointers parsed from the given {@code String}.
   * <p>Example:</p>
   * <pre><code>
   * convertToMem("ABC12345");
   * {"1", "2", "3", "28", "29", "30", "31", "32"}
   * </code></pre>
   * @param  in          the input {@code String} to be parsed
   * @param  memIndics   tells if the given {@code String} has memory indicators <code>${.*}</code> or <code>$.*</code>.
   * @param  labels      the map used to convert memory indicators into UFB memory pointers
   * @return             the {@code String[]} with all the UFB memory pointers
  */
  public static String[] convertToMem(
    final String in, final boolean memIndics, final HashMap<String, Integer> labels
  ){
    final String ANSI_RESET="\u001B[0m";
    final ArrayList<String> mems=new ArrayList<>();
    boolean backSlash=false;
    boolean memIndicator=false;
    boolean isLabel=false;
    final StringBuilder placeHolder=new StringBuilder();
    if(!memIndics){
      for(final char c:in.toCharArray()){
        if(memMap.containsKey(c))
          mems.add(memMap.get(c).toString());
        else{
          if(c=='\\'){
            if(backSlash){
              backSlash=false;
              mems.add("21");
              mems.add("21");
              for(
                final char c2:
                manPadding(Integer.toString('\\'), 4).toCharArray()
              )
                mems.add(memMap.get(c2).toString());
            }else{
              backSlash=true;
            }
          }else if(backSlash){
            if(c=='n'){
              mems.add("37");
            }else{
              mems.add("21");
              mems.add("21");
              for(
                final char c2:
                manPadding(Integer.toString((int)c), 4).toCharArray()
              )
                mems.add(memMap.get(c2).toString());
            }
            backSlash=false;
          }else{
            mems.add("21");
            mems.add("21");
            for(
              final char c2:
              manPadding(Integer.toString((int)c), 4).toCharArray()
            )
              mems.add(memMap.get(c2).toString());
          }
        }
      }
      return mems.toArray(new String[mems.size()]);
    }
    for(final char c:in.toCharArray()){
      if(c=='$'){
        memIndicator=true;
        placeHolder.append(c);
      }else if(memIndicator){
        placeHolder.append(c);
        if(c=='{')
          isLabel=true;
        else if(isLabel){
          if(c=='}'){
            final String key=placeHolder.substring(2, placeHolder.length()-1);
            if(labels.containsKey(key))
              mems.add(Integer.toString(labels.get(key)));
            else{
              System.out.printf("%s%s%s%s\n",
                ANSI_RESET, "\u001B[91m", formatError(
                  new String[]{convertUnicode(in)},
                  "Memory Index Label Already Replaced By Another",
                  placeHolder.toString(),
                  "Should Be Replaced With The Appropriate Label"
                ), ANSI_RESET
              );
              System.exit(1);
            }
            placeHolder.setLength(0);
            memIndicator=false;
            isLabel=false;
          }
        }else if(!isDigit(c)){
          memIndicator=false;
          for(final String converted:convertToMem(placeHolder.toString(), false, labels))
            mems.add(converted);
          placeHolder.setLength(0);
        }else{
          if(placeHolder.length()==4){
            mems.add(placeHolder.substring(1));
            placeHolder.setLength(0);
            memIndicator=false;
          }
        }
      }else if(memMap.containsKey(c))
        mems.add(memMap.get(c).toString());
      else{
        if(c=='\\'){
          if(backSlash){
            backSlash=false;
            mems.add("21");
            mems.add("21");
            for(
              final char c2:
              manPadding(Integer.toString('\\'), 4).toCharArray()
            )
              mems.add(memMap.get(c2).toString());
          }else{
            backSlash=true;
          }
        }else if(backSlash){
          if(c=='n'){
            mems.add("37");
          }else{
            mems.add("21");
            mems.add("21");
            for(
              final char c2:
              manPadding(Integer.toString((int)c), 4).toCharArray()
            )
              mems.add(memMap.get(c2).toString());
          }
          backSlash=false;
        }else{
          mems.add("21");
          mems.add("21");
          for(
            final char c2:
            manPadding(Integer.toString((int)c), 4).toCharArray()
          )
            mems.add(memMap.get(c2).toString());
        }
      }
    }
    return mems.toArray(new String[mems.size()]);
  }
}

File: src/src/universal.rs
use crate::memory_map::MemoryMap;

use std::collections::HashMap;
use std::env::consts::OS;

pub struct Universal {}

impl Universal {
    pub fn convert_u32_to_char(code: u32) -> char {
        char::from_u32(code).unwrap_or('\0')
    }

    pub fn arr_to_string<T: std::fmt::Display>(arr: &[T]) -> String {
        let mut out: String = String::new();
        arr.iter()
            .for_each(|x| out.push_str(&(x.to_string() + ",")));
        out[..out.len() - 2].to_string()
    }

    pub fn err_exit(err_msg: String) {
        if !OS.contains("windows") {
            println!("\u{001B}[91m{err_msg}\u{001B}[0m");
        } else {
            println!("{err_msg}");
        }
        std::process::exit(1);
    }

    pub fn manage_padding(input: String, padding: usize) -> String {
        format!("{:0>width$}", input, width = padding)
    }

    pub fn format_error(line: &[String], input: &[&str]) -> String {
        let arr: String = Self::arr_to_string(line).replace(", ", " ");
        format!(
            "Error: |\n    {}: |\n        \"{}\" {}: |\n            {arr}",
            input[0], input[1], input[2]
        )
    }

    pub fn convert_to_mem(
        input: &str,
        contains_labels: bool,
        labels: &HashMap<String, u8>,
        mem_map: &MemoryMap,
    ) -> Vec<String> {
        let mut out: Vec<String> = Vec::<String>::new();
        let mut back_slash: bool = false;
        let u: String = "21".to_string();
        if !contains_labels {
            for x in input.chars() {
                if mem_map.contains_key(&x.to_string()) {
                    out.push(mem_map.get(&x.to_string()).to_string());
                } else if x == '\\' {
                    if back_slash {
                        back_slash = false;
                        out.push(u.clone());
                        out.push(u.clone());
                        for x2 in Self::manage_padding(('\\' as u32).to_string(), 4).chars() {
                            out.push(mem_map.get(&x2.to_string()).to_string());
                        }
                    } else {
                        back_slash = true;
                    }
                } else if back_slash {
                    match x {
                        'n' => out.push("37".to_string()),
                        'r' => {
                            out.push(u.clone());
                            out.push(u.clone());
                            for x2 in "0032".chars() {
                                out.push(mem_map.get(&x2.to_string()).to_string());
                            }
                        }
                        'f' => {
                            out.push(u.clone());
                            out.push(u.clone());
                            for x2 in "0012".chars() {
                                out.push(mem_map.get(&x2.to_string()).to_string());
                            }
                        }
                        'b' => {
                            out.push(u.clone());
                            out.push(u.clone());
                            for x2 in "0008".chars() {
                                out.push(mem_map.get(&x2.to_string()).to_string());
                            }
                        }
                        _ => {
                            out.push(u.clone());
                            out.push(u.clone());
                            for x2 in Self::manage_padding((x as u32).to_string(), 4).chars() {
                                out.push(mem_map.get(&x2.to_string()).to_string());
                            }
                        }
                    }
                    back_slash = false;
                } else {
                    out.push(u.clone());
                    out.push(u.clone());
                    for x2 in Self::manage_padding((x as u32).to_string(), 4).chars() {
                        out.push(mem_map.get(&x2.to_string()).to_string());
                    }
                }
            }
            return out;
        };
        let mut mem_indicator: bool = false;
        let mut is_label: bool = false;
        let mut place_holder: String = String::new();
        for x in input.chars() {
            if x == '$' {
                mem_indicator = true;
                place_holder.push(x);
            } else if mem_indicator {
                place_holder.push(x);
                if x == '{' {
                    is_label = true;
                } else if is_label {
                    if x == '}' {
                        let key: String =
                            place_holder.clone()[2..place_holder.len() - 1].to_string();
                        if let Some(x) = labels.get(&key) {
                            out.push(x.to_string());
                        } else {
                            Self::err_exit(Self::format_error(
                                &[Self::convert_unicode(input)],
                                &[
                                    "Memory Index Label Already Replaced By Another",
                                    &place_holder,
                                    "Should Be Replaced With The Appropriate Label",
                                ],
                            ));
                        }
                        place_holder = String::new();
                        mem_indicator = false;
                        is_label = false;
                    }
                } else if !Self::is_digit(x) {
                    mem_indicator = false;
                    for converted in Self::convert_to_mem(&place_holder, false, labels, mem_map) {
                        out.push(converted);
                    }
                    place_holder = String::new();
                } else if place_holder.len() == 4 {
                    out.push(place_holder[1..].to_string());
                    place_holder = String::new();
                    mem_indicator = false;
                }
            } else if mem_map.contains_key(&x.to_string()) {
                out.push(mem_map.get(&x.to_string()).to_string());
            } else if x == '\\' {
                if back_slash {
                    back_slash = false;
                    out.push(u.clone());
                    out.push(u.clone());
                    for x2 in Self::manage_padding((x as u32).to_string(), 4).chars() {
                        out.push(mem_map.get(&x2.to_string()).to_string());
                    }
                } else {
                    back_slash = true;
                }
            } else if back_slash {
                if x == 'n' {
                    out.push("37".to_string());
                } else {
                    out.push(u.clone());
                    out.push(u.clone());
                    for x2 in Self::manage_padding((x as u32).to_string(), 4).chars() {
                        out.push(mem_map.get(&x2.to_string()).to_string());
                    }
                }
                back_slash = false;
            } else {
                out.push(u.clone());
                out.push(u.clone());
                for x2 in Self::manage_padding((x as u32).to_string(), 4).chars() {
                    out.push(mem_map.get(&x2.to_string()).to_string());
                }
            }
        }
        out
    }
    pub fn is_digit(c: char) -> bool {
        let val: u32 = c as u32;
        val > 47 && val < 58
    }
    pub fn quick_parse(input: String) -> u32 {
        let mut out: u32 = 0;
        input.chars().for_each(|x| {
            out += x as u32 - 48;
            out *= 10;
        });
        out / 10
    }
    pub fn convert_unicode(input: &str) -> String {
        if input.len() < 6 {
            return input.to_string();
        }
        let mut out: String = input.to_string();
        let mut lowercase_out: String = input.to_lowercase();
        while let Some(x) = lowercase_out.find("uu") {
            if x + 5 >= out.len() {
                break;
            }
            let mut place_holder: String = out[x + 2..x + 6].to_string();
            for x in place_holder.chars() {
                if !Self::is_digit(x) {
                    place_holder = String::new();
                    break;
                }
            }
            if !place_holder.is_empty() {
                {
                    let s1: String = out[..x].to_string();
                    let s2: String =
                        Self::convert_u32_to_char(Self::quick_parse(place_holder)).to_string();
                    let s3: String = out[x + 6..].to_string();
                    out.clear();
                    out.push_str(&(s1 + &s2 + &s3));
                }
                let s1: String = lowercase_out[..x].to_string();
                let s2: String = "_".to_string();
                let s3: String = lowercase_out[x + 6..].to_string();
                lowercase_out.clear();
                lowercase_out.push_str(&(s1 + &s2 + &s3));
            } else {
                let s1: String = lowercase_out[..x].to_string();
                let s2: String = "______".to_string();
                let s3: String = lowercase_out[x + 6..].to_string();
                lowercase_out.clear();
                lowercase_out.push_str(&(s1 + &s2 + &s3));
            }
        }
        out
    }
}
